extensions/ui/org.eclipse.smarthome.ui.basic/src/main/java/org/eclipse/smarthome/ui/basic/internal/render/AbstractWidgetRenderer.java

```diff
@@ -83,9 +83,11 @@ protected String preprocessSnippet(String snippet, Widget w) {
         snippet = StringUtils.replace(snippet, "%widget_id%", itemUIRegistry.getWidgetId(w));
         snippet = StringUtils.replace(snippet, "%icon_type%", config.getIconType());
         snippet = StringUtils.replace(snippet, "%item%", w.getItem() != null ? w.getItem() : "");
-        snippet = StringUtils.replace(snippet, "%label%", getLabel(w));
-        snippet = StringUtils.replace(snippet, "%value%", getValue(w));
-        snippet = StringUtils.replace(snippet, "%has_value%", new Boolean(hasValue(w)).toString());
+        // Optimization: avoid calling 3 times itemUIRegistry.getLabel(w)
```

Comment:
While I agree this makes sense, I also would have preferred to keep it separate.